File: app/src/main/java/com/lambdaschool/congressdata/CongresspersonProfile.java
package com.lambdaschool.congressdata;

import android.graphics.Bitmap;

import com.lambdaschool.congressdataapiaccess.Committee;
import com.lambdaschool.congressdataapiaccess.CongressDao;
import com.lambdaschool.congressdataapiaccess.CongresspersonDetails;
import com.lambdaschool.congressdataapiaccess.Subcommittee;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.ArrayList;
import java.util.List;

public class CongresspersonProfile {

    public static final String INDEPENDENT             = "Independent";
    public static final String REPUBLICAN              = "Republican";
    public static final String DEMOCRAT                = "Democrat";
    public static final String COMMITTEE_DETAIL_TARGET = "name";

    private String firstName, middleName, lastName, displayName, party, state, id, twitterAccount,
            facebookAccount, district, phone, location, office;
    private float missedVotesPct, votesWithPartyPct;
    private ArrayList<String> committees, subcommittees;
    private Bitmap image;

    public CongresspersonProfile() {
    }

    /**
     * This is a constructor
     *
     * @param firstName
     * @param middleName
     * @param lastName
     * @param party
     * @param state
     * @param id
     * @param twitterAccount
     * @param facebookAccount
     * @param district
     * @param phone
     * @param missedVotesPct
     * @param office
     * @param votesWithPartyPct
     */
    public CongresspersonProfile(String firstName, String middleName, String lastName, String party,
                                 String state, String id, String twitterAccount,
                                 String facebookAccount, String district, String phone,
                                 String missedVotesPct, String office, String votesWithPartyPct) {
        this.firstName         = firstName;
        this.middleName        = middleName;
        this.lastName          = lastName;
        this.state             = state;
        this.office            = office;
        this.district          = district;
        this.location          = state + " - " + "District " + district;
        this.id                = id;
        this.twitterAccount    = twitterAccount;
        this.facebookAccount   = facebookAccount;
        this.phone             = phone;
        this.missedVotesPct    = Float.parseFloat(missedVotesPct);
        this.votesWithPartyPct = Float.parseFloat(votesWithPartyPct);
        this.displayName       = buildDisplayName();

        switch (party) {
            case "R":
                this.party = REPUBLICAN;
                break;
            case "D":
                this.party = DEMOCRAT;
                break;
            case "I":
                this.party = INDEPENDENT;
                break;
        }
    }

    public CongresspersonProfile(CongresspersonDetails congresspersonDetails) {

            this(congresspersonDetails.getFirstName(),
                    congresspersonDetails.getMiddleName(),
                    congresspersonDetails.getLastName(),
                    congresspersonDetails.getCurrentParty(),
                    congresspersonDetails.getRoles().get(0).getState(),
                    congresspersonDetails.getMemberId(),
                    congresspersonDetails.getTwitterAccount(),
                    congresspersonDetails.getFacebookAccount(),
                    congresspersonDetails.getRoles().get(0).getDistrict(),
                    congresspersonDetails.getRoles().get(0).getPhone(),
                    Double.toString(congresspersonDetails.getRoles().get(0).getMissedVotesPct()),
                    congresspersonDetails.getRoles().get(0).getOffice(),
                    Double.toString(congresspersonDetails.getRoles().get(0).getVotesWithPartyPct()));

        //get correct role index
        int index = 0;
        for(int i = 0; i < congresspersonDetails.getRoles().size(); ++i) {
            if(congresspersonDetails.getRoles().get(i).getCongress().equals(CongressDao.CONGRESS_NUMBER)) {
                index = i;
                break;
            }
        }

        final List<Committee> committees = congresspersonDetails.getRoles().get(index).getCommittees();
        this.setCommittees(committees);
        final List<Subcommittee> subcommittees = congresspersonDetails.getRoles().get(index).getSubcommittees();
        this.setSubcommittees(subcommittees);
    }

    private String buildDisplayName() {
        StringBuilder nameBuilder = new StringBuilder();
        nameBuilder.append(firstName).append(" ");
        if (!middleName.equals("null")) {
            nameBuilder.append(middleName).append(" ");
        }
        nameBuilder.append(lastName);
        return nameBuilder.toString();
    }

    public float getPrimaryProgress() {
        // primary progress has been arbitrarily chosen to be red
        if(this.party.equals(REPUBLICAN)) {
            return this.votesWithPartyPct;
        } else {
            return 100 - this.votesWithPartyPct - this.missedVotesPct;
        }
    }

    public float getSecondaryProgress() {
        // secondary progress has been arbitrarily chosen to be blue
        if(this.party.equals(DEMOCRAT)) {
            return this.votesWithPartyPct;
        } else {
            return 100 - this.votesWithPartyPct - this.missedVotesPct;
        }
    }

    public ArrayList<String> getCommittees() {
        return committees;
    }

    public void setCommittees(JSONArray committeesJson) {
        this.committees = parseJsonList(committeesJson, COMMITTEE_DETAIL_TARGET);
    }

    public void setCommittees(List<Committee> committees) {
        this.committees = new ArrayList<>();
        for(Committee committee: committees) {
            this.committees.add(committee.getName());
        }
    }

    public ArrayList<String> getSubcommittees() {
        return subcommittees;
    }

    public void setSubcommittees(JSONArray subcommitteesJson) {
        this.subcommittees = this.committees = parseJsonList(subcommitteesJson, COMMITTEE_DETAIL_TARGET);;
    }

    public void setSubcommittees(List<Subcommittee> subcommittees) {
        this.subcommittees = new ArrayList<>();
        for(Subcommittee subcommittee: subcommittees) {
            this.subcommittees.add(subcommittee.getName());
        }
    }

    private ArrayList<String> parseJsonList(JSONArray array, String target) {
        ArrayList<String> result = new ArrayList<>();

        for(int i = 0; i < array.length(); ++i) {
            try {
                result.add(array.getJSONObject(i).getString(target));
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }

        return result;
    }

    public String getOffice() {
        return office;
    }

    public String getLocation() {
        return location;
    }

    public String getDisplayName() {
        return displayName;
    }

    public String getParty() {
        return party;
    }

    public String getState() {
        return state;
    }

    public String getId() {
        return id;
    }

    public String getTwitterAccount() {
        return twitterAccount;
    }

    public String getFacebookAccount() {
        return facebookAccount;
    }

    public String getDistrict() {
        return district;
    }

    public String getPhone() {
        return phone;
    }

    public float getMissedVotesPct() {
        return missedVotesPct;
    }

    public Bitmap getImage() {
        return image;
    }

    public void setImage(Bitmap image) {
        this.image = image;
    }
}

File: app/src/main/java/com/lambdaschool/congressdata/CongresspersonProfile.kt
package com.lambdaschool.congressdata

import android.graphics.Bitmap

import com.lambdaschool.congressdataapiaccess.Committee
import com.lambdaschool.congressdataapiaccess.CongressDao
import com.lambdaschool.congressdataapiaccess.CongresspersonDetails
import com.lambdaschool.congressdataapiaccess.Subcommittee

import org.json.JSONArray
import org.json.JSONException

import java.util.ArrayList

class CongresspersonProfile(
        val firstName: String,
        val middleName: String,
        val lastName: String,
        var party: String,
        val state: String,
        val id: String,
        val twitterAccount: String,
        val facebookAccount: String,
        val district: String,
        val phone: String, missedVotesPct: String,
        val office: String, votesWithPartyPct: String) {

    val displayName: String
    val location: String
    val missedVotesPct: Float
    private val votesWithPartyPct: Float
    var committees: ArrayList<String>? = null
        private set
    var subcommittees: ArrayList<String>? = null
        private set
    var image: Bitmap? = null

    // primary progress has been arbitrarily chosen to be red
    val primaryProgress: Float
        get() = if (this.party == REPUBLICAN) {
            this.votesWithPartyPct
        } else {
            100f - this.votesWithPartyPct - this.missedVotesPct
        }

    // secondary progress has been arbitrarily chosen to be blue
    val secondaryProgress: Float
        get() = if (this.party == DEMOCRAT) {
            this.votesWithPartyPct
        } else {
            100f - this.votesWithPartyPct - this.missedVotesPct
        }

    init {
        this.location = "$state - District $district"
        this.missedVotesPct = java.lang.Float.parseFloat(missedVotesPct)
        this.votesWithPartyPct = java.lang.Float.parseFloat(votesWithPartyPct)
        this.displayName = buildDisplayName()
        when (party) {
            "R" -> this.party = REPUBLICAN
            "D" -> this.party = DEMOCRAT
            "I" -> this.party = INDEPENDENT
        }
    }

    constructor(congresspersonDetails: CongresspersonDetails) : this(congresspersonDetails.firstName,
            congresspersonDetails.middleName,
            congresspersonDetails.lastName,
            congresspersonDetails.currentParty,
            congresspersonDetails.roles[0].state,
            congresspersonDetails.memberId,
            congresspersonDetails.twitterAccount,
            congresspersonDetails.facebookAccount,
            congresspersonDetails.roles[0].district,
            congresspersonDetails.roles[0].phone,
            java.lang.Double.toString(congresspersonDetails.roles[0].missedVotesPct),
            congresspersonDetails.roles[0].office,
            java.lang.Double.toString(congresspersonDetails.roles[0].votesWithPartyPct)) {

        //get correct role index
        var index = 0
        for (i in 0 until congresspersonDetails.roles.size) {
            if (congresspersonDetails.roles[i].congress == CongressDao.CONGRESS_NUMBER) {
                index = i
                break
            }
        }

        val committees = congresspersonDetails.roles[index].committees
        this.setCommittees(committees)
        val subcommittees = congresspersonDetails.roles[index].subcommittees
        this.setSubcommittees(subcommittees)
    }

    private fun buildDisplayName(): String {
        val nameBuilder = StringBuilder()
        nameBuilder.append(firstName).append(" ")
        if (middleName != "null") {
            nameBuilder.append(middleName).append(" ")
        }
        nameBuilder.append(lastName)
        return nameBuilder.toString()
    }

    fun setCommittees(committeesJson: JSONArray) {
        this.committees = parseJsonList(committeesJson, COMMITTEE_DETAIL_TARGET)
    }

    fun setCommittees(committees: List<Committee>) {
        this.committees = ArrayList()
        for (committee in committees) {
            this.committees!!.add(committee.name)
        }
    }

    fun setSubcommittees(subcommitteesJson: JSONArray) {
        this.committees = parseJsonList(subcommitteesJson, COMMITTEE_DETAIL_TARGET)
        this.subcommittees = this.committees
    }

    fun setSubcommittees(subcommittees: List<Subcommittee>) {
        this.subcommittees = ArrayList()
        for (subcommittee in subcommittees) {
            this.subcommittees!!.add(subcommittee.name)
        }
    }

    private fun parseJsonList(array: JSONArray, target: String): ArrayList<String> {
        val result = ArrayList<String>()

        for (i in 0 until array.length()) {
            try {
                result.add(array.getJSONObject(i).getString(target))
            } catch (e: JSONException) {
                e.printStackTrace()
            }

        }

        return result
    }

    companion object {

        val INDEPENDENT = "Independent"
        val REPUBLICAN = "Republican"
        val DEMOCRAT = "Democrat"
        val COMMITTEE_DETAIL_TARGET = "name"
    }
}
